src/compiler/checker.ts

```diff
@@ -34307,7 +34331,12 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return resolveErrorCall(node);
         }
 
-        return resolveCall(node, callSignatures, candidatesOutArray, checkMode, callChainFlags);
+        // If the call expression is a synthetic call to a `[Symbol.hasInstance]` method then we will produce a head
+        // message when reporting diagnostics that explains how we got to `right[Symbol.hasInstance](left)` from
+        // `left instanceof right`, as it pertains to "Argument" related messages reported for the call.
+        return resolveCall(node, callSignatures, candidatesOutArray, checkMode, callChainFlags, () => isSyntheticHasInstanceMethodCall(node) ?

```

Comment:
is there another way to conditionally issue this error? I guess you'd have to check each time you issue the error, but I would guess it would only have to be in a couple of places, not the number of places that had to change to make this a callback in the current commit.

Reply:
`resolveCall` already allowed you to pass in a `headMessage`, which was the way you would conditionally choose to append this message. I changed this to a callback so as to lazily compute the head message rather than do eager, expensive work up front when there may be no error to report.

I could pass in an _additional_ `boolean` parameter that moves this computation to the point before we issue an error, or add a bitflag to `CheckMode` possibly.

Reply:
Actually, I don't need an extra parameter. I can just perform the check prior to reporting errors in absence of a more specific `headMessage`.